Publish fsx in nuget

realmarv opened the pull request.

knocte commented on .github/workflows/CI.yml:
```diff
@@ -445,7 +445,7 @@ jobs:
       run: make check
     - name: compile this repo's .fsx scripts with fsx
-      run: ./compileFSharpScripts.fsx
+      run: sudo ./compileFSharpScripts.fsx
```
@realmarv are we sure these sudos are needed now? let's do a commit removing them to see what happens

realmarv replied:
Done

knocte commented on fsx-legacy.sln:
```diff
@@ -52,6 +52,7 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "build", "build", "{DBAE02AB
 	EndProjectSection
 EndProject
 Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "Fsdk.Tests", "Fsdk.Tests\Fsdk.Tests-legacy.fsproj", "{43BA7E25-975B-4DF9-B274-EEF6C806C1D0}"
+Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "fsx-legacy", "fsx\fsx-legacy.fsproj", "{14E47DE0-49B3-4B58-9627-8C3F10A21D24}"
```
the name can stay as `fsx`, no need to be `fsx-legacy`

realmarv replied:
Done

knocte commented on fsx.sln:
```diff
@@ -44,15 +44,16 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "test", "test", "{C51659D1-9
+		Tools\safeRun.fsx = Tools\safeRun.fsx
+		Tools\nugetPush.fsx = Tools\nugetPush.fsx
+		Tools\fsi.bat = Tools\fsi.bat
+		Tools\gitPush1by1.fsx = Tools\gitPush1by1.fsx
```
@realmarv this hunk hints me that you're not using `git gui` tool to do your commits, right?

knocte replied:
> But overall yesterday I wasn't completely done with cleaning the code

Fair enough. But:

> Yes.

Then use `git gui` please, it's a time saver to catch these things early on, not include it in the commit, and save 5min of back&forth in the PR :)

realmarv replied:
Ok

knocte commented on fsx/Program.fs:
```diff
@@ -134,13 +131,29 @@ let InjectBinSubfolderInPath(userScript: FileInfo) =
-
-let sourceDir = DirectoryInfo __SOURCE_DIRECTORY__
+let thisScriptFileName =
+    ((new FileInfo(
```
@realmarv in F#, the keyword `new` is only required when you're creating a new object of a type that implements IDisposable, which is not the case here, so you can remove it

realmarv replied:
Done

knocte commented on fsx/Program.fs:
```diff
@@ -134,13 +131,29 @@ let InjectBinSubfolderInPath(userScript: FileInfo) =
+
+// if thisScriptFileName <> "fsx.dll" then
+//     failwith
+//         "this launcher should have been renamed to fsx.dll at install time; please report this bug"
```
@realmarv seems you forgot again 1 bullet from our Workflow Guidelines ;) https://github.com/nblockchain/conventions/blob/master/WorkflowGuidelines.md

realmarv replied:
Done

knocte commented on fsx/Program.fs:
```diff
@@ -134,13 +131,29 @@ let InjectBinSubfolderInPath(userScript: FileInfo) =
+//         "this launcher should have been renamed to fsx.dll at install time; please report this bug"
+
+let sourceDir =
+    ((new FileInfo(
```
@realmarv same here about `new`

realmarv replied:
Done

knocte commented on fsx/Program.fs:
```diff
@@ -134,13 +131,29 @@ let InjectBinSubfolderInPath(userScript: FileInfo) =
+            .GetExecutingAssembly()
+            .Location
+    ))
+        .Directory) //DirectoryInfo __SOURCE_DIRECTORY__
```
@realmarv same here about comment

realmarv replied:
Done

knocte commented on fsx/Program.fs:
```diff
@@ -150,8 +163,11 @@ let fsxcAssembly =
+    failwith(
+        sprintf
+            "fsxc assembly not found in %A; please report this bug"
+            fsxcAssembly.FullName
```
I guess you meant `fsxcAssembly.Directory.FullName`, otherwise the message is a tiny bit confusing

realmarv replied:
Done

knocte commented on scripts/make.fsx:
```diff
@@ -223,13 +252,42 @@ match maybeTarget with
+        )
     )
 
+    if assemblyExecutableExtension = "dll" then
```
@realmarv it's more readable to have `#if !LEGACY_FRAMEWORK`

realmarv replied:
Done

knocte commented on scripts/make.fsx:
```diff
@@ -223,13 +252,42 @@ match maybeTarget with
+                "bin",
+                "Release",
+                "net6.0",
+                "fsx.runtimeconfig.dev.json"
```
@realmarv are you sure copying both json files is needed? I bet this one is not really needed, let's try removing it to see if CI is still green

realmarv replied:
Done

knocte commented on scripts/make.fsx:
```diff
@@ -195,6 +195,35 @@ let fsxBatDestination =
+        sprintf "fsx.%s" assemblyExecutableExtension
+    )
+    |> FileInfo
+#endif
```
doing this would be shorter and have less duplication:

```
    let fsxLauncher =
        Path.Combine(
            RootDir.FullName,
            "fsx",
            "bin",
            "Release"
#if !LEGACY_FRAMEWORK
            , "net6.0"
#endif
            , sprintf "fsx.%s" assemblyExecutableExtension
        ) |> FileInfo
```

realmarv replied:
Done

knocte commented:
Last but not least, I would rename the title to "Convert launcher.fsx to a project". After addressing all my feedback, we would be much closer to finishing this task, good work so far.

knocte commented:
> I would rename the title to "Convert launcher.fsx to a project"

Please note, I didn't say "PR title".

BTW, I saw you pushed 1 by 1, and I don't understand why. If you're addressing my feedback in different commits, that's fine, but as soon as you see CI green on them, there's no point to not squash everything into 1. Therefore, pushing 1 by 1 when you're going to squash later, doesn't make much sense (it just clogs your Actions tab of so many unnecessary CI jobs).

knocte commented:
> I didn't want to wait 20 mins to apply the next comment 🤔

By clogging your github account with so many unnecessary jobs, you caused an even bigger wait than 20min in my opinion.
BTW the issue is not so much about pushing 1 by 1, the thing I would have not done, myself, would be to do `git rebase -i` if you know already that these commits will be squashed with the 1st one (as they don't have any reason to leave separated from the 1st commit).

knocte commented on .github/workflows/CI.yml:
```diff
@@ -485,7 +485,6 @@ jobs:
       run: |
         # to clean Debug artifacts first (make install builds in Release config)
         git clean -fdx
-
```
@realmarv this hunk is not needed

realmarv replied:
Done

realmarv replied:
Done

knocte commented on fsx/Program.fs:
```diff
@@ -134,13 +131,30 @@ let InjectBinSubfolderInPath(userScript: FileInfo) =
+            .GetExecutingAssembly()
+            .Location
+    )
+        .Directory
```
@realmarv you're storing a directory in a variable called "thisScriptFileName"

realmarv replied:
Done

knocte commented on scripts/make.fsx:
```diff
@@ -223,13 +243,28 @@ match maybeTarget with
+            "net6.0",
+            "fsx.runtimeconfig.json"
+        ),
+        Path.Combine(fsxInstallationDir.FullName, "fsx.runtimeconfig.json")
```
@realmarv I'm glad that fsx.runtimeconfig.dev.json was not needed; now I wonder if this one above is needed either? let's remove it and see if CI is still green

realmarv replied:
It broke the CI

knocte replied:
wow, very interesting

realmarv commented:
> > I didn't want to wait 20 mins to apply the next comment thinking
> 
> By clogging your github account with so many unnecessary jobs, you caused an even bigger wait than 20min in my opinion. BTW the issue is not so much about pushing 1 by 1, the thing I would have not done, myself, would be to do `git rebase -i` if you know already that these commits will be squashed with the 1st one (as they don't have any reason to leave separated from the 1st commit).

Right

knocte commented:
Let's also remove this line from the ReadMe file as part of the commit of this PR: https://github.com/nblockchain/fsx/blame/master/ReadMe.md#L10

knocte commented:
After you have addressed all the feedback, this PR needs to get rebased. (But please squash to 1 commit before rebasing; this way the `git rebase` command will potentially pose less conflicts to deal with, if any.)

knocte commented on fsx/Program.fs:
```diff
@@ -141,10 +141,9 @@ let thisScriptFileName =
-// if thisScriptFileName <> "fsx.dll" then
-//     failwith
-//         "this launcher should have been renamed to fsx.dll at install time; please report this bug"
+if thisScriptFileName <> "fsx.dll" then
```
@realmarv I bet this will fail, don't you remember that the extension is different dependending of LEGACY_FRAMEWORK? ;)

realmarv replied:
Right

knocte commented on fsx/Program.fs:
```diff
@@ -134,13 +131,29 @@ let InjectBinSubfolderInPath(userScript: FileInfo) =
+            .Reflection
+            .Assembly
+            .GetExecutingAssembly()
+            .Location
```
@realmarv you're retreiving this location twice, let's not have duplicated code and save it first in a variable yeah?

realmarv replied:
Right

realmarv commented:
> Let's also remove this line from the ReadMe file as part of the commit of this PR: https://github.com/nblockchain/fsx/blame/master/ReadMe.md#L10

Done

realmarv commented:
> After you have addressed all the feedback, this PR needs to get rebased. (But please squash to 1 commit before rebasing; this way the `git rebase` command will potentially pose less conflicts to deal with, if any.)

Done

knocte commented:
Looking great! Now, let's finish the original task by renaming the PR to "Publish fsx in nuget", and have a 2nd commit (do not squash with 1st), that publishes the fsx.fsproj to nuget.org in the same way you did with fsxc.fsproj.

knocte commented:
Ok good, the 2nd commit might have looked easy enough to do, but I'm afraid that it will not be as easy as you think. Why do I say this? Because I'm sure that, if published as is, it would fail. The reason is that fsx has a dependency on fsxc, but currently that dependency is a runtime-dependency. That is, it fsx tries to look for fsxc.dll (or fsxc.exe) in the same folder where fsx.dll (or fsx.exe) is located. This won't work when being installed as dotnet nuget tool. Even if you add a `<PackageReference>` to fsx.fsproj, I doubt it would work either, so I'm afraid that you would need to do the following instead:
*. First, rename fsxc/Program.fs to fsxc/Fsxc.fs.
* Include Fsxc.fs in fsx.fsproj and fsx-legacy.fsproj (but as links, in the same way you did with the Fsdk files).
* Try to use fsxc functions from fsx/Program.fs (instead of invoking fsxc as an executable). For this, you might need to change the API of Fsxc.fs slightly for your needs.

knocte commented:
Last but not least, when you have the above finished and working (CI green), please remove this line from the ReadMe in the 2nd commit: https://github.com/nblockchain/fsx/blame/master/ReadMe.md#L9

knocte commented on ReadMe.md:
```diff
@@ -7,7 +7,6 @@ THIS REPO IS FACING A COMPLETE OVERHAUL/REVAMP/RENOVATION IN ORDER TO SUPPORT .N
 Unfinished tasks so far:
 * Revamp this ReadMe.md file to remove any mentions to Mono or the legacy .NET4.x framework.
 * To reach a v1.0 release, publish fsx as dotnet tool in nuget.
```
@realmarv removing the line above should be done in 2nd commit ^

realmarv replied:
Done

knocte commented on fsxc/Fsxc.fs:
```diff
@@ -0,0 +1,945 @@
+module Program =
+
+#if LEGACY_FRAMEWORK
+    let public nugetExeTmpLocation: Lazy<FileInfo> =
```
@realmarv whenever you need to increase the visibility of an API to fix a problem, please don't go all the way to the other extreme without testing first something in the middle: internal

realmarv replied:
Done

knocte commented on scripts/make.fsx:
```diff
@@ -227,13 +247,28 @@ match maybeTarget with
+            "fsx",
+            "bin",
+            "Release",
+            "net6.0",
```
@realmarv this path `Path.Combine(RootDir.FullName, "fsx", "bin", "Release", "net6.0")` has been used before in this file, so you could save it in a variable to avoid repeating yourself

realmarv replied:
Done

knocte commented:
This is almost ready! good work, don't despair, will review couple of things more.

knocte commented on fsx/Program.fs:
```diff
@@ -6,20 +6,26 @@ open System.Text
 open System.Diagnostics
 
-#r "System.Configuration"
 open System.Configuration
```
@realmarv I think this line above can be removed too?

realmarv replied:
Done

knocte commented on fsx/Program.fs:
```diff
@@ -150,34 +157,25 @@ let fsxcAssembly =
-        Arguments = sprintf "%s %s" (String.Join(" ", fsxcArgs)) userScript
-    }
+let fsxcMainArguemnts = List.toArray ((Seq.toList fsxcArgs) @ [userScript])
+
```
@realmarv before your modification, fsxcCmd was different in LEGACY_FRAMEWORK vs !LEGACY_FRAMEWORK; but after your modification, both lines look the same, so you can remove the `#if`

realmarv replied:
Done

knocte commented on fsx/Program.fs:
```diff
@@ -150,34 +157,25 @@ let fsxcAssembly =
 
-let proc = Process.Execute(fsxcCmd, Echo.Off)
-proc.UnwrapDefault() |> ignore<string>
+Program.Main(fsxcMainArguemnts) |> ignore
```
@realmarv unneeded parentheses

realmarv replied:
Done

knocte commented on fsx/Program.fs:
```diff
@@ -150,34 +157,25 @@ let fsxcAssembly =
-                (String.Join(" ", fsxcArgs))
-                userScript
-    }
+let fsxcMainArguemnts = List.toArray ((Seq.toList fsxcArgs) @ [userScript])
```
@realmarv this line is very unreadable; you start from fsxcArgs which is a seq, you convert it to a list, then you create a second list from userScript, and then you append both lists, and then you convert the result to an array! It can be simplified greatly, for example, by doing: `let fsxcMainArguments = Seq.append fsxcArgs (Seq.exactlyOne userScript) |> Seq.toArray` (btw notice the typo `fsxcMainArguemnts` -> `fsxcMainArguments`

realmarv replied:
Done

knocte commented on fsx/Program.fs:
```diff
@@ -133,14 +130,23 @@ let InjectBinSubfolderInPath(userScript: FileInfo) =
+if thisScriptFileName <> (sprintf "fsx.%s" assemblyExecutableExtension) then
+    failwith(
+        sprintf
+            "this launcher should have been renamed to fsx.%s at install time; please report this bug"
```
@realmarv think about the original message for a second; it existed because the file `launcher.fsx` ended up in the user's /usr/bin as `fsx.fsx`. After this commit, however, this is not the case anymore. First because there's no rename anymore, it's now a project compilation into a DLL/EXE, not a script rename. So this message needs be reworked (or even removed entirely, not sure)

realmarv replied:
Done

knocte commented on fsx/Program.fs:
```diff
@@ -150,34 +155,20 @@ let fsxcAssembly =
 if not fsxcAssembly.Exists then
-    failwith
-        "fsxc assembly not found in the same folder as this launcher; please report this bug"
+    failwith(
```
@realmarv the second commit of this PR doesn't call fsxc assembly anymore under the hood, so this failwith becomes obsolete, it has to be removed on the second commit

realmarv replied:
Done

knocte commented on fsxc/Program.fs:
```diff
@@ -0,0 +1,13 @@
+        Program.Main argv
+    finally
+#if LEGACY_FRAMEWORK
+        if Program.nugetExeTmpLocation.IsValueCreated then
```
by placing this logic outside Fsxc.fs, then this logic is not happening when a script is compiled by fsx, what would be the reason for that?

realmarv replied:
Not sure. Should I create a function inside Fsxc.fs with the content of the current main function and call that in the main function?

knocte replied:
I guess that would be better than the current state of the PR

realmarv replied:
I just put `try finally` in the Main function

knocte replied:
that is not acceptable, as explained in another comment

realmarv replied:
Ok changed it

knocte commented on scripts/make.fsx:
```diff
@@ -227,13 +249,21 @@ match maybeTarget with
 
+#if !LEGACY_FRAMEWORK
+    File.Copy(
+        Path.Combine(finalReleaseFolderPath, "fsx.runtimeconfig.json"),
```
@realmarv the commit that introduces this change is a bit confusing, I mean, a developer that reads it in the future would not really understand why this is being done; to prevent that, please state in the commit message the error that we were getting when this was not being done

realmarv replied:
Done

realmarv commented on fsxc/Program.fs:
```diff
@@ -0,0 +1,5 @@
+
+[<EntryPoint>]
+let main argv =
+    Program.Main argv
```
Done

knocte replied:
You had to fix this in both places, not just one.

knocte replied:
and please make Main private, in order to avoid this happening again

realmarv replied:
Done

knocte commented on fsxc/Fsxc.fs:
```diff
@@ -45,7 +45,7 @@ exception NoScriptProvided
 
 #if LEGACY_FRAMEWORK
-    let private nugetExeTmpLocation: Lazy<FileInfo> =
+    let internal nugetExeTmpLocation: Lazy<FileInfo> =
```
@realmarv I think this doesn't need to change now

realmarv replied:
Done

knocte commented on fsx/Program.fs:
```diff
@@ -133,14 +129,14 @@ let InjectBinSubfolderInPath(userScript: FileInfo) =
-        "this launcher should have been renamed to fsx.fsx at install time; please report this bug"
-
-let sourceDir = DirectoryInfo __SOURCE_DIRECTORY__
+let sourceDir = FileInfo(assemblyLocation).Directory
```
@realmarv these 2 vars above are not being used and can be removed (in 2nd commit?)

realmarv replied:
Done

knocte commented on fsx/Program.fs:
```diff
@@ -133,14 +129,14 @@ let InjectBinSubfolderInPath(userScript: FileInfo) =
-let sourceDir = DirectoryInfo __SOURCE_DIRECTORY__
+let sourceDir = FileInfo(assemblyLocation).Directory
 
 let fsxcAssembly =
```
@realmarv same here, unused so can be removed

realmarv replied:
Done

knocte commented on scripts/make.fsx:
```diff
@@ -195,6 +195,28 @@ let fsxBatDestination =
+let finalReleaseFolderPath = releaseFolderPath
+#endif
+
+let fsxLauncher =
```
@realmarv cosmetic: let's move this down to where it was before (in line 230)

realmarv replied:
Done
